### ISSUE TITLE
How about tag_invoke?

### ISSUE THREAD

**Issue by seekstar** [serde-cpp/serde-cpp#5]
According to the code of `serde_gen`, the implementation of serialize and deserialize of custom class seems to be based on struct template specialization. However, the specialized struct template should be defined in `namespace serde`, which is impossible for private classes. I didn't test it because I don't know exactly how to use this library and how to check the generated C++ files. Another issue of struct template specialization is that it is possible to specialize `serde::Serialize<ClassInLibA>` in two different libraries, e.g., library B and library C. Then if we link library B and library C at the same time, it will produce symbol conflicts.

I propose to use `tag_invoke`. It only reserves one name `tag_invoke` in the user namespace, and it solves all problems described above. Its behavior is very similar to Rust's trait. For example, with `tag_invoke`, you can either implement `serialize` for a known type along with the definition of the `serialize` interface (e.g., implement `serialize` for std types in this library) or implement `serialize` for a custom type along with the definition of that type. In another word, it does not allow implementing third-party interfaces for third-party classes. Although it seems to be somewhat restrictive, it eliminates ambiguousness. I have tried to use `tag_invoke` to implement serialization and deserialization, and it works fine: <https://github.com/iidb/wing/blob/9e1ceabea133af976c966b73b0b88ec525a99dfd/src/common/serde.hpp>

**Comment by natanaeljr** [comment]
You are right, there is currently no way to support serialization generation for private classes, although, there is a way to still use the serde APIs explicitly if you implement `serialize` and `deserialize` methods in that private class. See example test case [Advanced.LocalPrivateTypes](https://github.com/serde-cpp/serde-cpp/blob/main/serde-cpp/serde_yaml/test/advanced.cpp#L55).

Regarding usage docs/examples, I plan on creating a separate example repository to serve as a template project that uses serde-cpp. That is to show how to use serde-cpp cmake package and call the `serde_generate_target` function from CMake to generate the serialization/deserialization files.

Your `tag_invoke` suggestion seems interesting, I will look into your linked example and evaluate how to bring it to serde-cpp. We may continue discussing viability and implementation details in this thread.
